src/os/bluestore/BlueStore.cc

```diff
@@ -6322,11 +6323,19 @@ void BlueStore::_txc_state_proc(TransContext *txc)
 		   << dendl;
 	} else {
 	  _txc_finalize_kv(txc, txc->t);
+	  if (g_conf->bluestore_rtc && txc->wal_txn && !txc->no_rtc) {
+	    if (txc->osr->kv_finisher_submitting == 1) {
+	      db->submit_transaction_sync(txc->t);
+	      _txc_state_proc(txc);
+	      return;
```

Comment:
It's not redundant check.
Without checking txc->wal_txn, ShardedOpWQ tries to take osr->qlock, multiple times.